SDK location not found

## 异常现象

首次克隆项目，`./gradlew build` 失败：
```
Caused by: com.android.builder.errors.EvalIssueException: SDK location not found. Define location with an ANDROID_SDK_ROOT environment variable or by setting the sdk.dir path in your project's local properties file at '/Users/bob/AndroidProjects/Shadow/projects/sdk/coding/local.properties'.
        at com.android.builder.errors.IssueReporter.reportError(IssueReporter.kt:109)
        at com.android.builder.errors.IssueReporter.reportError$default(IssueReporter.kt:105)
        at com.android.build.gradle.internal.SdkLocator.getSdkLocation(SdkLocator.kt:239)
        at com.android.build.gradle.internal.SdkLocator.getSdkDirectory(SdkLocator.kt:197)
        at com.android.build.gradle.internal.SdkComponentsKt.getSdkDir(SdkComponents.kt:368)
        at com.android.build.gradle.internal.SdkComponentsBuildService$VersionedSdkLoader$sdkDirectoryProvider$1.call(SdkComponents.kt:198)
        at com.android.build.gradle.internal.SdkComponentsBuildService$VersionedSdkLoader$sdkDirectoryProvider$1.call(SdkComponents.kt:197)
        at org.gradle.api.internal.provider.DefaultProvider.calculateOwnValue(DefaultProvider.java:66)
        at org.gradle.api.internal.provider.AbstractMinimalProvider.calculateValue(AbstractMinimalProvider.java:103)
        at org.gradle.api.internal.provider.TransformBackedProvider.calculateOwnValue(TransformBackedProvider.java:64)
        at org.gradle.api.internal.provider.AbstractMinimalProvider.calculateValue(AbstractMinimalProvider.java:103)
        at org.gradle.api.internal.provider.MappingProvider.calculateOwnValue(MappingProvider.java:55)
        at org.gradle.api.internal.provider.AbstractMinimalProvider.calculateValue(AbstractMinimalProvider.java:103)
        at org.gradle.api.internal.provider.DefaultProperty.calculateValueFrom(DefaultProperty.java:128)
        at org.gradle.api.internal.provider.DefaultProperty.calculateValueFrom(DefaultProperty.java:25)
        at org.gradle.api.internal.provider.AbstractProperty.doCalculateValue(AbstractProperty.java:133)
        at org.gradle.api.internal.provider.AbstractProperty.calculateOwnValue(AbstractProperty.java:127)
        at org.gradle.api.internal.provider.AbstractMinimalProvider.calculateValue(AbstractMinimalProvider.java:103)
        at org.gradle.api.internal.provider.FlatMapProvider.calculateOwnValue(FlatMapProvider.java:50)
        at org.gradle.api.internal.provider.AbstractMinimalProvider.get(AbstractMinimalProvider.java:84)
        at org.gradle.api.provider.Provider$get.call(Unknown Source)
        at build_6980bvnws4asrriy8vrvrlha.run(/Users/bob/AndroidProjects/Shadow/projects/sdk/coding/get-android-jar/build.gradle:9)
```

环境：macOS 11.1(15款macbook pro)，Android Studio Arctic Fox | 2020.3.1 patch 3，openJDK 11，Gradle 7.0.2，Android Gradle Plugin 7.0.3

## 相关资料

[issue #355](https://github.com/Tencent/Shadow/issues/355) 

[patch](https://github.com/Tencent/Shadow/commit/f80ece3a5fcfd5e2fc0b4d7049ee797b5cfd2fbb)

Android Gradle Plugin 7.0.2 [SDKLocator.kt](https://github.com/Tencent/Shadow/files/7638592/SDKLocator.txt)

## 原因分析

最早 Shadow 项目依赖环境变量${ANDROID_HOME}，后改为 Shadow/projects/sdk/coding/get-android-jar module 提供 android.jar，通过 Shadow/projects/sdk/coding/get-android-jar/build.gradle 动态获取 Android SDK 位置的方式来找到 android.jar：
```groovy
def sdkPath = android.sdkComponents.getSdkDirectory().get().asFile.absolutePath
def androidJarPath = new File(sdkPath, "platforms/${android.compileSdkVersion}/android.jar")
```

但是，通过查看 AGP 7.0.2 代码发现最终调用获取的还是 local.properties 定义的 sdk.dir 或者 ${ANDROID_SDK_ROOT} ${ANDROID_HOME} 之类的环境变量，一旦这些方式无法获取 SDK location，就会默认赋值 missingSdkDirectory：
```
> Configure project :coding:get-android-jar
==> sdkPath=/Users/bob/AndroidProjects/Shadow/projects/sdk/coding/missingSdkDirectory
==> androidJarPath=/Users/bob/AndroidProjects/Shadow/projects/sdk/coding/missingSdkDirectory/platforms/android-30/android.jar
```

也就是说，这一改动其实是没有任何意义的，最终还是回到了依赖环境变量的方式（local.properties的方式肯定不能考虑，无法保证每个人的 SDK 位置一样），反而少了原来的环境变量检测告警。

## 复现及解决

分析完原因，复现也很简单，只需要把环境变量  ${ANDROID_SDK_ROOT} ${ANDROID_HOME}  之类给移除 并删除 module 的  local.propertie 后，build 一定会失败。

本地解决方式也简单，配置环境变量  ${ANDROID_SDK_ROOT} ${ANDROID_HOME}  任一即可。

对于Shadow项目来说，在没有更好的方式前建议还是回滚到原来的方案。事实上，以后应该也不会有更好方案，每个人本地的 SDK 位置不一样，AGP 也只能通过用户配置的 SDK 位置来获取，要么配置在文件(local.properties) 要么配置在环境变量。
如果不回滚方案，也可以考虑 readme 补回相关环境变量说明。

--- a/README.md
+++ b/README.md
@@ -30,25 +30,10 @@ Shadow不仅开源分享了插件技术的关键代码，还完整的分享了�
 ## 编译与开发环境
 
 ### 环境准备
-第一次clone Shadow的代码到本地后，建议先在命令行编译一次。
+建议直接用最新的稳定版本Android Studio打开工程。目前项目已适配`Android Studio Arctic Fox | 2020.3.1`，
+低版本的Android Studio可能因为Gradle版本过高而无法正常打开项目。
 
-* 在编译时，**必须**使用`gradlew`脚本，以保证采用了项目配置的Gradle版本。
-
-在命令行测试编译时可以执行这个任务：
-```
-./gradlew build
-```
-
-如果Gradle提示需要Java 11，建议直接使用Android Studio Arctic Fox内置的JDK。
-macOS：`export JAVA_HOME=/Applications/Android\ Studio.app/Contents/jre/Contents/Home`
-其他OS欢迎PR补充。
-
-如果没有出错，再尝试用Android Studio打开工程。
-
-* **必须**使用3.5或更高版本的Android Studio打开工程。(业务插件开发时没有限制)
-* **必须**关闭Android Studio的**Instant Run**功能。
-
-然后就可以在IDE中选择`sample-host`模块直接运行了。
+然后在IDE中选择`sample-app`或`sample-host`模块直接运行，分别体验同一份代码在正常安装情况下和插件情况下的运行情况。
 
 ![选择sample-host直接运行](pics/run-sample-host-in-ide.png)
 

--- a/settings.gradle
+++ b/settings.gradle
@@ -1,5 +1,8 @@
+import java.nio.file.Files
+
 rootProject.name = 'shadow'
 
+copyLocalPropertiesForIncludeBuilds()
 includeBuild 'projects/sdk/coding'
 includeBuild 'projects/sdk/core'
 includeBuild 'projects/sdk/dynamic'
@@ -77,4 +80,22 @@ def includeSampleDynamicApk() {
     project(':sample-hello-api-holder').projectDir = file('projects/sample/dynamic-apk/sample-hello-api-holder')
     project(':sample-hello-apk').projectDir = file('projects/sample/dynamic-apk/sample-hello-apk')
     project(':sample-hello-host').projectDir = file('projects/sample/dynamic-apk/sample-hello-host')
+}
+
+/**
+ * Android Studio当前不会为IncludeBuild创建包含sdk.dir的local.properties
+ * 使得没有ANDROID_HOME或等效环境变量时仅依赖根目录的local.properties无法编译IncludeBuild。
+ * 为了使此含有IncludeBuild的项目和其他不含有IncludeBuild的普通Android工程一样可以在
+ * 只有根目录的local.properties情况下正常编译，用此任务复制local.properties。
+ */
+def copyLocalPropertiesForIncludeBuilds() {
+    def rootFile = file('local.properties')
+    if (rootFile.exists()) {
+        ['coding', 'core', 'dynamic'].forEach {
+            def includeBuildFile = file("projects/sdk/${it}/local.properties")
+            if (!includeBuildFile.exists()) {
+                Files.copy(rootFile.toPath(), includeBuildFile.toPath())
+            }
+        }
+    }
 }

